nano/node/socket.cpp

```diff
@@ -238,55 +238,110 @@ void nano::server_socket::on_connection (std::function<bool (std::shared_ptr<nan
-					{
-						if (!ec_a)
-						{
-							// Make sure the new connection doesn't idle. Note that in most cases, the callback is going to start
-							// an IO operation immediately, which will start a timer.
-							new_connection->checkup ();
-							new_connection->start_timer (this_l->node.network_params.network.is_dev_network () ? std::chrono::seconds (2) : this_l->node.network_params.node.idle_timeout);
-							this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_success, nano::stat::dir::in);
-							this_l->connections.push_back (new_connection);
-						}
-						else
-						{
-							this_l->node.logger.try_log ("Unable to accept connection: ", ec_a.message ());
-						}
+				this_l->node.logger.always_log ("Network: max_inbound_connections reached, unable to open new connection");
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+				boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
+					this_l->on_connection (callback_a);
+				}));
+				return;
+			}
 
-						// If the callback returns true, keep accepting new connections
-						if (callback_a (new_connection, ec_a))
-						{
-							this_l->on_connection (callback_a);
-						}
-						else
-						{
-							this_l->node.logger.try_log ("Stopping to accept connections");
-						}
-					}
-					else
-					{
-						this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
-						boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
-							this_l->on_connection (callback_a);
-						}));
-					}
+			if (!ec_a)
+			{
+				// Make sure the new connection doesn't idle. Note that in most cases, the callback is going to start
+				// an IO operation immediately, which will start a timer.
+				new_connection->checkup ();
+				new_connection->start_timer (this_l->node.network_params.network.is_dev_network () ? std::chrono::seconds (2) : this_l->node.network_params.node.idle_timeout);
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_success, nano::stat::dir::in);
+				this_l->connections.push_back (new_connection);
+				callback_a (new_connection, ec_a);
+				this_l->on_connection (callback_a);
+				return;
+			}
+
+			this_l->node.logger.always_log ("Network: Unable to accept connection: ", ec_a.message ());
+			if (this_l->is_temporary_error (ec_a))
+			{
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+				boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
+					this_l->on_connection (callback_a);
 				}));
+				return;
 			}
-		}
+
+			if (this_l->is_system_error (ec_a))
+			{
+				boost::this_thread::sleep (boost::posix_time::milliseconds (500));
```

Comment:
If we should sleep, need to make sure what is the proper way do to it from within a strand. V21.3 did it using a async_wait, when needed. Good if someone with more experience in this code can recommend the best way. 

This also applies to the case in the max incoming connection limit check (where I think we should sleep a bit as well).